File: ios/bitchat/Services/SonarPushRegistration.swift
//
// SonarPushRegistration.swift
// bitchat
//
// Registers the device's APNS token with both notification servers:
//   1. Transponder — MIP-05 encrypted token shares (chat/call wakeups)
//   2. Breez NDS   — webhook URL (wallet wakeups, silent only)
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

#if os(iOS)

import Foundation
import os
import CryptoKit
import SonarCore

final class SonarPushRegistration {
    static let shared = SonarPushRegistration()

    private static let log = Logger(
        subsystem: Bundle.main.bundleIdentifier ?? "sh.hedwig.sonar",
        category: "SonarPush"
    )

    private var cachedAPNSToken: Data?
    private var cachedFCMToken: String?
    private var cachedOffer: String?
    private var sonarNode: SonarNode?
    private let queue = DispatchQueue(label: "chat.bitchat.sonar.push.registration")

    /// Persisted only for diagnostics; do not use it as a cross-launch skip.
    /// Boltz owns the authoritative offer webhook state, and a stale local marker
    /// can otherwise suppress the unregister -> register self-heal.
    private static let webhookMarkerKey = "breez_webhook_marker"
    private static let webhookMarkerVersion = "ios-fcm-explicit-token-v2"
    private var sessionWebhookMarker: String?

    private var transponderNpub: String {
        Bundle.main.infoDictionary?["TRANSPONDER_NPUB"] as? String ?? ""
    }

    /// Base URL of the Breez NDS, e.g. `https://nds.sonar.hedwig.sh`.
    ///
    /// The value is configured in `Local.xcconfig` as a BARE HOST (no scheme),
    /// because xcconfig treats `//` as a comment and silently truncates
    /// `https://host` to `https:`. We prepend the scheme here and reject the
    /// truncated `https:`/`http:` sentinels so a malformed webhook URL can never
    /// be registered with Boltz (which would leave offline receive broken).
    private var ndsUrl: String {
        let raw = (Bundle.main.infoDictionary?["NDS_URL"] as? String ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty, raw != "https:", raw != "http:" else { return "" }
        if raw.hasPrefix("http://") || raw.hasPrefix("https://") { return raw }
        return "https://\(raw)"
    }

    private init() {}

    // MARK: - Public

    func didRegisterForRemoteNotifications(deviceToken: Data) {
        let hex = deviceToken.map { String(format: "%02x", $0) }.joined()
        Self.log.info("APNS token collected (\(hex.prefix(8))...)")
        queue.async { self.cachedAPNSToken = deviceToken }
        // Transponder (chat/calls) uses the raw APNs token. The Breez NDS uses the
        // Firebase FCM token instead — registered via didReceiveFCMToken.
        registerTransponder(token: deviceToken)
    }

    /// Firebase FCM token for the Breez NDS webhook. `breez/notify` is FCM-only;
    /// Firebase bridges FCM → APNs on iOS (the APNs key lives in the Firebase
    /// Console). Matches Sonar Android and Unify.
    func didReceiveFCMToken(_ fcmToken: String, wallet: WalletBridgeService?) {
        Self.log.info("FCM token collected (\(fcmToken.prefix(8))...)")
        // Coordinate offer + token on the serial queue: the webhook can only be
        // (re)subscribed once both are known (the offer keys the idempotency
        // marker). Whichever lands second on the queue triggers the subscribe.
        queue.async {
            self.cachedFCMToken = fcmToken
            guard let offer = self.cachedOffer else {
                Self.log.info("Breez NDS: FCM token ready, waiting for receive offer")
                return
            }
            self.subscribeBreezWebhook(offer: offer, fcmToken: fcmToken, wallet: wallet)
        }
    }

    /// Called from the receiver-advertising flow each time the BOLT12 receive
    /// offer is (re)published. Coordinates the offer with the FCM token — whichever
    /// arrives second triggers the subscribe.
    func ensureBreezWebhook(offer: String, wallet: WalletBridgeService?) {
        queue.async {
            self.cachedOffer = offer
            guard let fcmToken = self.cachedFCMToken else {
                Self.log.info("Breez NDS: receive offer ready, waiting for FCM token")
                return
            }
            self.subscribeBreezWebhook(offer: offer, fcmToken: fcmToken, wallet: wallet)
        }
    }

    func retryBreezWebhookIfNeeded(wallet: WalletBridgeService) {
        queue.async {
            guard let fcmToken = self.cachedFCMToken, let offer = self.cachedOffer else { return }
            self.subscribeBreezWebhook(offer: offer, fcmToken: fcmToken, wallet: wallet)
        }
    }

    func unregister(wallet: WalletBridgeService? = nil) {
        queue.async {
            self.cachedAPNSToken = nil
            self.cachedOffer = nil
            self.sessionWebhookMarker = nil
        }
        // Force a fresh subscribe next time (e.g. after a wallet/seed change).
        UserDefaults.standard.removeObject(forKey: Self.webhookMarkerKey)
        if let wallet {
            Task { @MainActor in
                try? await wallet.unregisterWebhook()
            }
        }
        Self.log.info("Unregistered from push servers")
    }

    /// Set the SonarNode once it's initialized. Retries transponder registration
    /// if an APNS token was already collected.
    func setSonarNode(_ node: SonarNode) {
        queue.async {
            self.sonarNode = node
            if let token = self.cachedAPNSToken {
                self.registerTransponder(token: token)
            }
        }
    }

    // MARK: - Private

    private static let maxRetries = 3

    private func registerTransponder(token: Data) {
        guard !transponderNpub.isEmpty else { return }
        guard let node = sonarNode else {
            Self.log.info("Transponder: SonarNode not ready, will retry after setSonarNode")
            return
        }
        let npub = self.transponderNpub
        DispatchQueue.global(qos: .utility).async {
            var backoff: UInt32 = 2
            for attempt in 1...Self.maxRetries {
                do {
                    try node.registerPushToken(
                        platform: "apns",
                        token: token,
                        serverNpub: npub
                    )
                    Self.log.info("Transponder: MIP-05 push token registered")
                    return
                } catch {
                    Self.log.warning("Transponder registration attempt \(attempt)/\(Self.maxRetries) failed: \(error)")
                    if attempt < Self.maxRetries { sleep(backoff) }
                    backoff *= 2
                }
            }
        }
    }

    private func subscribeBreezWebhook(offer: String, fcmToken: String, wallet: WalletBridgeService?) {
        guard let webhookUrl = Self.webhookUrl(ndsUrl: ndsUrl, platform: "ios", fcmToken: fcmToken) else {
            Self.log.warning("Breez NDS disabled: invalid NDS_URL build setting")
            return
        }
        guard let wallet else {
            Self.log.info("Breez NDS: wallet not ready, will retry after wallet setup")
            return
        }
        let marker = Self.webhookMarker(offer: offer, webhookUrl: webhookUrl)
        if sessionWebhookMarker == marker {
            Self.log.info("Breez NDS webhook re-subscribe already queued for current offer")
            return
        }
        sessionWebhookMarker = marker
        Task { @MainActor in
            guard case .ready = wallet.state else {
                Self.log.info("Breez NDS: wallet not ready, will retry after wallet setup")
                self.queue.async {
                    if self.sessionWebhookMarker == marker {
                        self.sessionWebhookMarker = nil
                    }
                }
                return
            }
            do {
                // Force a clean re-subscribe. A plain registerWebhook no-ops when the
                // SDK's local `bolt12_offers.webhook_url` already equals this URL but
                // Boltz still holds `url=None` — the desync that breaks offline receive.
                // unregister clears the local webhook state, register re-PATCHes the
                // offer on Boltz (`update_bolt12_offer`) with the current FCM URL.
                // Matches the misty-breez reference's unregister→register flow.
                try? await wallet.unregisterWebhook()
                try await wallet.registerWebhook(url: webhookUrl)
                UserDefaults.standard.set(marker, forKey: Self.webhookMarkerKey)
                Self.log.info("Breez NDS webhook force re-subscribed for current offer (FCM)")
            } catch {
                self.queue.async {
                    if self.sessionWebhookMarker == marker {
                        self.sessionWebhookMarker = nil
                    }
                }
                Self.log.warning("Breez NDS webhook registration failed: \(error)")
            }
        }
    }

    private static func webhookUrl(ndsUrl: String, platform: String, fcmToken: String) -> String? {
        let raw = ndsUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: raw),
              let scheme = components.scheme?.lowercased(),
              scheme == "https" || scheme == "http",
              components.host?.isEmpty == false
        else { return nil }

        var path = components.path
        if path.hasSuffix("/") {
            path.removeLast()
        }
        components.path = "\(path)/api/v1/notify"
        var queryItems = components.queryItems ?? []
        queryItems.append(URLQueryItem(name: "platform", value: platform))
        queryItems.append(URLQueryItem(name: "token", value: fcmToken))
        components.queryItems = queryItems
        return components.url?.absoluteString
    }

    private static func webhookMarker(offer: String, webhookUrl: String) -> String {
        let digest = SHA256.hash(data: Data("\(webhookMarkerVersion)|\(offer)|\(webhookUrl)".utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

#endif

File: apps/sonar/composeApp/src/androidMain/kotlin/chat/bitchat/sonar/push/SonarPushRegistration.kt
package chat.bitchat.sonar.push

import android.net.Uri
import android.util.Log
import chat.bitchat.sonar.AppContextHolder
import chat.bitchat.sonar.BuildConfig
import chat.bitchat.sonar.SonarCore
import chat.bitchat.sonar.wallet.WalletBridge
import chat.bitchat.sonar.wallet.WalletState
import com.google.firebase.messaging.FirebaseMessaging
import java.security.MessageDigest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Registers the device's FCM token with both notification servers:
 *   1. Transponder — MIP-05 encrypted token shares (chat/call wakeups)
 *   2. Breez NDS — webhook URL (wallet wakeups, silent only)
 */
object SonarPushRegistration {

    private const val TAG = "SonarPush"
    private const val MAX_RETRIES = 3
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val transponderNpub: String get() = BuildConfig.TRANSPONDER_NPUB

    // Base URL of the Breez NDS. Tolerates a bare host (prepends https://) and
    // rejects the truncated "https:"/"http:" sentinels, mirroring iOS, so a
    // malformed webhook URL can never be registered (which breaks offline pay).
    private val ndsUrl: String
        get() {
            val raw = BuildConfig.NDS_URL.trim()
            if (raw.isEmpty() || raw == "https:" || raw == "http:") return ""
            return if (raw.startsWith("http://") || raw.startsWith("https://")) raw else "https://$raw"
        }

    @Volatile private var cachedFcmToken: String? = null
    @Volatile private var cachedOffer: String? = null
    @Volatile private var sessionWebhookMarker: String? = null

    fun ensureRegistered() {
        if (!SonarPushPrefs.effectivePushEnabled(AppContextHolder.ctx)) {
            Log.d(TAG, "Push not registered: disabled by user preference")
            return
        }
        if (transponderNpub.isBlank() && ndsUrl.isBlank()) {
            Log.d(TAG, "Push not configured (no TRANSPONDER_NPUB or NDS_URL)")
            return
        }
        FirebaseMessaging.getInstance().token.addOnSuccessListener { token ->
            Log.d(TAG, "FCM token collected")
            cachedFcmToken = token
            registerTransponder(token)
            retryBreezWebhookIfNeeded()
        }.addOnFailureListener { e ->
            Log.w(TAG, "FCM token collection failed", e)
        }
    }

    fun onTokenRefresh(token: String) {
        if (!SonarPushPrefs.effectivePushEnabled(AppContextHolder.ctx)) {
            Log.d(TAG, "FCM token refreshed while push disabled")
            cachedFcmToken = null
            return
        }
        Log.d(TAG, "FCM token refreshed")
        cachedFcmToken = token
        registerTransponder(token)
        retryBreezWebhookIfNeeded()
    }

    /**
     * The Breez webhook is offer-scoped on the swap server. Registering only the
     * FCM token can leave an existing BOLT12 offer with a stale/missing webhook,
     * so coordinate token + current offer and force a re-subscribe when either
     * changes. Mirrors the iOS `unregisterWebhook` -> `registerWebhook` path,
     * without trusting a persisted local marker across app launches.
     */
    fun ensureBreezWebhook(offer: String) {
        cachedOffer = offer
        retryBreezWebhookIfNeeded()
    }

    /**
     * Retry Breez webhook registration after the wallet becomes Ready.
     * Called from the wallet setup path once the SDK connects.
     */
    fun retryBreezWebhookIfNeeded() {
        val token = cachedFcmToken
        val offer = cachedOffer
        if (token == null) {
            Log.d(TAG, "Breez NDS: receive offer ready, waiting for FCM token")
            return
        }
        if (offer == null) {
            Log.d(TAG, "Breez NDS: FCM token ready, waiting for receive offer")
            return
        }
        registerBreezWebhook(token, offer)
    }

    private fun registerTransponder(fcmToken: String) {
        if (transponderNpub.isBlank()) return
        scope.launch {
            var backoff = 2_000L
            for (attempt in 1..MAX_RETRIES) {
                try {
                    SonarCore.start()
                    SonarCore.registerPushToken(
                        platform = "fcm",
                        token = fcmToken.toByteArray(Charsets.UTF_8),
                        serverNpub = transponderNpub,
                    )
                    Log.d(TAG, "Transponder: MIP-05 push token registered")
                    return@launch
                } catch (e: Exception) {
                    Log.w(TAG, "Transponder registration attempt $attempt/$MAX_RETRIES failed", e)
                    if (attempt < MAX_RETRIES) delay(backoff)
                    backoff *= 2
                }
            }
        }
    }

    private fun registerBreezWebhook(fcmToken: String, offer: String) {
        if (ndsUrl.isBlank()) return
        if (WalletBridge.state() !is WalletState.Ready) {
            Log.d(TAG, "Breez NDS: wallet not ready, will retry after wallet setup")
            return
        }
        val webhookUrl = webhookUrl(fcmToken)
        if (webhookUrl == null) {
            Log.w(TAG, "Breez NDS disabled: invalid NDS_URL build setting")
            return
        }
        val marker = webhookMarker(offer, webhookUrl)
        if (sessionWebhookMarker == marker) {
            Log.d(TAG, "Breez NDS webhook re-subscribe already queued for current offer")
            return
        }
        sessionWebhookMarker = marker
        scope.launch {
            try {
                try { WalletBridge.unregisterWebhook() } catch (_: Exception) {}
                WalletBridge.registerWebhook(webhookUrl)
                Log.d(TAG, "Breez NDS webhook force re-subscribed for current offer")
            } catch (e: Exception) {
                if (sessionWebhookMarker == marker) sessionWebhookMarker = null
                Log.w(TAG, "Breez NDS webhook registration failed", e)
            }
        }
    }

    fun unregister() {
        scope.launch {
            try { WalletBridge.unregisterWebhook() } catch (_: Exception) {}
        }
        cachedFcmToken = null
        cachedOffer = null
        sessionWebhookMarker = null
        Log.d(TAG, "Unregistered from push servers")
    }

    private fun webhookUrl(fcmToken: String): String? {
        val uri = runCatching { Uri.parse(ndsUrl.trim()) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase()
        if ((scheme != "https" && scheme != "http") || uri.host.isNullOrBlank()) return null
        return uri.buildUpon()
            .appendPath("api")
            .appendPath("v1")
            .appendPath("notify")
            .appendQueryParameter("platform", "android")
            .appendQueryParameter("token", fcmToken)
            .build()
            .toString()
    }

    private fun webhookMarker(offer: String, webhookUrl: String): String {
        val bytes = MessageDigest.getInstance("SHA-256")
            .digest("$offer|$webhookUrl".toByteArray(Charsets.UTF_8))
        return bytes.joinToString(separator = "") { "%02x".format(it.toInt() and 0xff) }
    }
}
